edgexfoundry/edgex-ui-go
# build: automate angular test, build, and dev tasks

The main goal of this PR is to have Jenkins run the Angular unit tests during the test stage, but getting that working is most of the way to making build/dev tasks work with Docker, so I've included those where it made sense:

- Added a Dockerfile and Makefile to the `web/` directory to do the following within a container:
  - run the Angular unit tests
  - build production versions of the site for each locale (output to `web/dist/`)
  - run the dev server with hot-reloading
- Updated the `web/README.md` to fix the Angular version number and add info about the new Makefile/Dockerized build process
- Updated the root Makefile: 
  - run the web unit tests upon `make test`
  - add a `web` target to copy the production build to the `static` directory
- Configured karma so that unit tests can run in a container on kernels that don't have userns enabled
- Update the `angular.json` to reduce development build time and produce localized builds (note: it looks like there are a lot of missing translations)
- Checked in an updated `package-lock.json` file

As an aside, it seems the versions of things in the web project are pretty behind the latest versions -- probably pretty typical in the JavaScript world. I found that the project simply will not build the newest version of Node, hence why this is tied to a version a couple releases behind. When I built it the first time, it updated the `package-lock.json`, generating a HUGE diff compared to the original, and flagged several critical vulnerabilities.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] I have added unit tests for the new feature or bug fix (if not, why?): N/A here
- [ ] I have fully tested (add details below) this the new feature or bug fix (if not, why?): N/A
- [ ] I have opened a PR for the related docs change (if not, why?): because they belong with this PR

## Testing Instructions
This assumes you have Docker and Make installed:

- clone the fork
- run the following from the repo root:
  - `make test`: should test _both_ the Go code _and_ the Angular unit tests
  - `make test-go` and `make test-angular` to test individually
  - `make web` to build the production website & copy it to the `static` directory
- run the following from the `web` directory:
  - `make test`: runs the Angular unit tests
  - `make build`: creates a production build
  - `make serve`: runs the dev server, but see notes in the README about potential networking issues and Angular CLI signal handling
